### ISSUE TITLE
Proposal: Automatic `__sig` generation for JS library functions

### ISSUE THREAD

**Issue by sbc100** [emscripten-core/emscripten#18979]
We use `__sig` attributes on JS library symbols for 2 different purposes:

1. When adding JS functions to the wasm table at runtime using `addFunction` (this will fail unless `.sig` attribute is present)
2. Tagging pointer args and return values when building with `-sMEMORY64`

Adding these signatures by hand is error prone and incremental.   We tend to do reactively as we find gaps.

I'd like to propose a system where the majority of these `__sig` attributes could be auto-generated.   The proposal would involve creating new tools that we would run at SDK build time.    Something like `gen_sig_info.py` analogous to the existing `gen_struct_info.py`.

== Part 1 - Generate a list of all JS symbols ==

This is not as simple as it sounds because the list of JS symbols often depends on the settings used at link time.  I imagine that due to this issue we might be able to auto-generate a list that covers all symbols in a single build, but we can start with a best effort approach.

Its likely we can't get 100% symbol coverage from the single build.   This means we might still need explict `__sig` attributes for some symbols, but since there would be very few of these we could make that lack of `__sig` into an error and easily catch when we need to add them.

We should also make it into an error to specify a `__sig` when there exists and auto-generated one.

== Part 2 - Build a C program that references every JS symbols ==

This is also not as simple as it initially sounds since not all JS symbols exist in headers.  For some JS symbols their only native declaration is deep in some `.c` file that uses it.   I proposal that to solve this problem we require that all JS symbol are declared in a internal header somewhere (e.g. `threading_iternal.h`),

== Part 3 - Extract signature information from the compiled program ==

There are several ways to achieve this, but the two most logical ones seem to be:

a. Analyze the llvm bitcode.  This would clearly distinguish pointer vs non-pointer, but might be a little complex (e.g. I'm not sure I want to write an IR parser in python, or a new LLVM IR pass).

b. Analyze the resulting wasm module.  We already have python code to extract signature information.   The tricky part is that pointer vs non-pointer information is lost.  We could compile once with wasm32 and again with wasm64 and look for places where we see i32 vs i64 differences and assume pointers?  That would also apply to `long` params, but don't know of any JS library funcs that use `long`.. perhaps we could ban the use of `long` in our core JS library code?

**Comment by sbc100** [comment]
@dschuff @kripken WDTY?

**Comment by sbc100** [comment]
Regarding 3b, we are not building an entire user program, just a single C source file (like `gen_struct_info.py`).   Also, like `gen_struct_info.py`, the result can be cached.

The way we get the sigs is by compiling the program from `Part 2`.  Just like `gen_struct_info.py` it would involve an auto-generated single-source-file program that might look like this:

```
/** Auto-generated by gen_sig_info.py **/
#include <all-include-files>  // Important that all JS symbols live in these headers.

void* all_addresses = [
  &emscripten_get_now,
  &emscripten_xxx,
  &emscripten_yyy,
  ....
];

int main() {
  // This program does nothing, it just has to keep all addresses alive
  return (int)all_addresses;
}
```

Compiling this program should be very fast.

**Comment by sbc100** [comment]
(not that in order to create the above source code we don't need any signature information because all function symbols can be cast to void*)

**Comment by sbc100** [comment]
@juj too

**Comment by sbc100** [comment]
Right, but size_t and pointers are both considerd "pointer-sized".. i.e. we use `p` for both of them.   So we want to detect size_t as pointer-sized.

IIUC its only long that would be a problem, and using long in a JS library function is asking for trouble for this reason (I think).

Actually I worry that if we go with the LLVM bitcode approach we don't be able to distinguish size_t from long maybe?

**Comment by sbc100** [comment]
Starting work on this has already uncovered a bunch of dead code: #18984 #18983

**Comment by brendandahl** [comment]
To add:
`__sig` is also used with JSPI to create the Function wrappers.

Another alternative would be a clang plugin, [here's an example](https://github.com/llvm/llvm-project/blob/main/clang/examples/PrintFunctionNames/PrintFunctionNames.cpp) of a simple one that prints all the functions in a file.

**Comment by sbc100** [comment]
Good point regarding the clang plugin.  All add that as an option (c).

So far I have had good success with simply analyzing the wasm output.
